### PR TITLE
[Quantization] Add support for fused rowwise-quantizing SLWS

### PR DESCRIPTION
**Pull request by jfix71**
*Description*: Previously rowwise-quantized SLWS was only enabled by directly loading it as quantized. This allows for rowwise-quantizing it through our normal profile + quantization flow.

*Testing*: Added two unit tests; tested on other private models.

*Documentation*: Updated.

### CODE REVIEW

File: lib/Quantization/Quantization.cpp [hunk 1]
```diff
@@ -476,14 +476,19 @@ class FunctionQuantizer : public FunctionConverter {
     lastMorphedNodeWithTypeChanges = nullptr;
   }
 
-  /// Traverse all nodes to find quantized FullyConnected nodes, and convert it
-  /// to RowwiseQuantizedFullyConnected if the weights is constant.
+  /// Traverse all nodes to find applicable quantized nodes, and convert them
+  /// to RowwiseQuantized versions if applicable is constant.
```
Comment (rdzhabarov) [inline lib/Quantization/Quantization.cpp]: `if applicable is constant`? Did you mean something else here?

File: lib/Quantization/Quantization.cpp [hunk 2]
```diff
@@ -502,31 +507,68 @@ class FunctionQuantizer : public FunctionConverter {
...
+            // Replace the usages of quantized FC node to
+            // RowwiseQuantizedFullyConnected Node.
+            result.replaceAllUsesOfWith(fcq->getResult());
           }
         }
       }
+
+      // Convert SLWS from normal quantized version to fused rowwise-quantized
+      // version if applicable. Data must be Constant for this to occur. We also
+      // will not quantized the weights as we do for the default normal
```
Comment (rdzhabarov) [inline lib/Quantization/Quantization.cpp]: typo `quantized.

File: tests/unittests/QuantizationTest.cpp [hunk 3]
```diff
@@ -333,6 +333,58 @@ TEST(Quantization, enableRowwiseQuantizedFullyConnected) {
...
+
+  auto *data = mod.createPlaceholder(ElemKind::FloatTy, {3, 1}, "data", false);
+  auto *weights =
+      mod.createPlaceholder(ElemKind::FloatTy, {8}, "weights", false);
+  auto *indices =
+      mod.createPlaceholder(ElemKind::Int64ITy, {8}, "indices", false);
+  auto *lengths =
+      mod.createPlaceholder(ElemKind::Int32ITy, {4}, "lengths", false);
+
+  bindings.allocate(data)->init(Tensor::InitKind::Xavier, 3, mod.getPRNG());
```
Comment (rdzhabarov) [inline tests/unittests/QuantizationTest.cpp]: nit: we do not necessarily need to `init` this data (we already kinda profile it by having `std::vector<NodeQuantizationInfo>`.

File: lib/Quantization/Quantization.cpp [hunk 4]
```diff
@@ -502,31 +507,68 @@ class FunctionQuantizer : public FunctionConverter {
...
+      // quantized SLWS, as the rowwise version uses float weights.
+      if (auto *SLWS =
+              llvm::dyn_cast<SparseLengthsWeightedSumNode>(Q->getInput())) {
+        NodeValue data = SLWS->getData();
+        auto *dataQN = llvm::dyn_cast<QuantizeNode>(data.getNode());
+        assert(dataQN && "Data input should have been quantized");
```
Comment (rdzhabarov) [inline lib/Quantization/Quantization.cpp]: any reason those checks are asserts vs for FC it's actually part of the precondition before conversion?

Comment (jfix71) [reply lib/Quantization/Quantization.cpp]: Hmm, well I guess it's possible that the FC was not quantized due to `doNotQuantizeKinds`. I'll make it a precondition, thanks!

**Pull request merged**